File: packages/agent/daemon/claudesidecar/sessionsettings.go
package claudesidecar

import (
	"os"
	"sync/atomic"
)

// SessionSettings mirrors SidecarSessionSettings.
type SessionSettings struct {
	Model            string
	PermissionModeID string
	PlanMode         bool
	Effort           string
	Speed            string
}

// ConfigOption mirrors SidecarConfigOption.
type ConfigOption struct {
	ID           string             `json:"id"`
	Name         string             `json:"name,omitempty"`
	Description  string             `json:"description,omitempty"`
	Category     string             `json:"category,omitempty"`
	Type         string             `json:"type,omitempty"`
	CurrentValue string             `json:"currentValue,omitempty"`
	Options      []ConfigOptionItem `json:"options"`
}

// ConfigOptionItem is one selectable value of a ConfigOption.
type ConfigOptionItem struct {
	Value       string `json:"value"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
}

// PendingFlagSettings carries queued live-settings changes for
// apply_flag_settings. A nil pointer field is "unset"; effortLevel may be an
// explicit null on the wire, mirrored by NullEffort.
type PendingFlagSettings struct {
	EffortLevel  string
	EffortSet    bool
	EffortIsNull bool
	FastMode     bool
	FastModeSet  bool
}

func (p PendingFlagSettings) isEmpty() bool {
	return !p.EffortSet && !p.FastModeSet
}

func (p PendingFlagSettings) toWire() map[string]any {
	settings := map[string]any{}
	if p.EffortSet {
		if p.EffortIsNull {
			settings["effortLevel"] = nil
		} else {
			settings["effortLevel"] = p.EffortLevel
		}
	}
	if p.FastModeSet {
		settings["fastMode"] = p.FastMode
	}
	return settings
}

func sidecarSessionSettings(payload map[string]any) *SessionSettings {
	settings := recordValue(payload["settings"])
	permissionModeID := firstNonEmptyText(
		stringValue(payload["permissionModeId"]),
		stringValue(settings["permissionModeId"]),
	)
	if permissionModeID == "" {
		permissionModeID = "default"
	}
	effort := firstNonEmptyText(
		stringValue(payload["effort"]),
		stringValue(settings["effort"]),
		stringValue(settings["reasoningEffort"]),
	)
	return &SessionSettings{
		Model:            stringValue(settings["model"]),
		PermissionModeID: permissionModeID,
		PlanMode:         booleanValue(settings["planMode"]),
		Effort:           effort,
		Speed:            stringValue(settings["speed"]),
	}
}

func effectivePermissionMode(settings *SessionSettings) string {
	if settings.PlanMode {
		return "plan"
	}
	permissionMode := permissionModeValue(settings.PermissionModeID)
	if permissionMode == "bypassPermissions" && !canBypassPermissions() {
		return "default"
	}
	return permissionMode
}

func permissionModeValue(value string) string {
	switch value {
	case "default", "acceptEdits", "bypassPermissions", "plan", "dontAsk", "auto":
		return value
	default:
		return ""
	}
}

func modelOptionValue(value string) string {
	model := normalizeTitle(value)
	if model == "" || model == "default" {
		return ""
	}
	return model
}

func sidecarModelOptionsFromInitializationResult(value map[string]any) []ConfigOptionItem {
	rawModels, _ := value["models"].([]any)
	options := make([]ConfigOptionItem, 0, len(rawModels))
	seen := map[string]struct{}{}
	for _, item := range rawModels {
		model := recordValue(item)
		if model == nil {
			continue
		}
		modelValue := firstNonEmptyText(
			stringValue(model["value"]),
			stringValue(model["id"]),
			stringValue(model["modelId"]),
			stringValue(model["model_id"]),
		)
		if modelValue == "" {
			continue
		}
		if _, duplicate := seen[modelValue]; duplicate {
			continue
		}
		seen[modelValue] = struct{}{}
		name := firstNonEmptyText(
			stringValue(model["displayName"]),
			stringValue(model["display_name"]),
			stringValue(model["name"]),
			modelValue,
		)
		options = append(options, ConfigOptionItem{
			Value:       modelValue,
			Name:        name,
			Description: stringValue(model["description"]),
		})
	}
	return options
}

func defaultSidecarModelOptionValue(options []ConfigOptionItem) string {
	for _, option := range options {
		if option.Value == "default" {
			return option.Value
		}
	}
	if len(options) > 0 {
		return options[0].Value
	}
	return "default"
}

func effortLevelValue(value string) (string, bool) {
	switch value {
	case "low", "medium", "high", "xhigh":
		return value, true
	default:
		return "", false
	}
}

func flagSettingsFromSessionSettings(settings *SessionSettings) PendingFlagSettings {
	result := PendingFlagSettings{}
	if settings.Effort != "" {
		level, valid := effortLevelValue(settings.Effort)
		result.EffortSet = true
		if valid {
			result.EffortLevel = level
		} else {
			result.EffortIsNull = true
		}
	}
	switch settings.Speed {
	case "fast":
		result.FastModeSet = true
		result.FastMode = true
	case "standard":
		result.FastModeSet = true
		result.FastMode = false
	}
	return result
}

func querySettingsFromSessionSettings(settings *SessionSettings) map[string]any {
	result := map[string]any{}
	switch settings.Speed {
	case "fast":
		result["fastMode"] = true
	case "standard":
		result["fastMode"] = false
	}
	return result
}

func approvalOptions() []map[string]any {
	return []map[string]any{
		{"kind": "allow_always", "name": "Allow for session", "optionId": "allow_always"},
		{"kind": "allow_once", "name": "Allow", "optionId": "allow"},
		{"kind": "reject_once", "name": "Reject", "optionId": "reject"},
	}
}

func exitPlanOptions() []map[string]any {
	options := []map[string]any{
		{"kind": "allow_always", "name": `Yes, and use "auto" mode`, "optionId": "auto"},
		{"kind": "allow_always", "name": "Yes, and auto-accept edits", "optionId": "acceptEdits"},
		{"kind": "allow_once", "name": "Yes, and manually approve edits", "optionId": "default"},
		{"kind": "reject_once", "name": "No, keep planning", "optionId": "plan"},
	}
	if canBypassPermissions() {
		options = append([]map[string]any{
			{"kind": "allow_always", "name": "Yes, and bypass permissions", "optionId": "bypassPermissions"},
		}, options...)
	}
	return options
}

func isAllowOption(optionID string) bool {
	switch optionID {
	case "allow", "allow_always", "accept", "acceptEdits", "default", "auto", "bypassPermissions":
		return true
	default:
		return false
	}
}

func isExitPlanAllowOption(optionID string) bool {
	if optionID == "bypassPermissions" {
		return canBypassPermissions()
	}
	switch optionID {
	case "default", "acceptEdits", "auto":
		return true
	default:
		return false
	}
}

// sandboxed marks the sidecar as daemon-embedded. The daemon always spawned
// the TypeScript sidecar with IS_SANDBOX=1; the in-process Go sidecar sets
// this instead so bypass-permission behavior stays identical.
var sandboxed atomic.Bool

// SetSandboxed marks the process as sandboxed the way IS_SANDBOX=1 does.
func SetSandboxed(value bool) {
	sandboxed.Store(value)
}

func canBypassPermissions() bool {
	isRoot := os.Geteuid() == 0
	return !isRoot || sandboxed.Load() || os.Getenv("IS_SANDBOX") != ""
}

File: packages/agent/claude-sdk-sidecar/src/sessionSettings.ts
import type {
  EffortLevel,
  PermissionMode,
  Settings
} from "@anthropic-ai/claude-agent-sdk";
import { recordValue } from "./normalizer.ts";
import { booleanValue, stringValue } from "./runtimeValues.ts";

const claudePlansDirectoryEnv = "TUTTI_CLAUDE_PLANS_DIRECTORY";

/**
 * Pending live flag-settings write. Mirrors the SDK's own `applyFlagSettings`
 * parameter shape: `effortLevel` accepts the full {@link EffortLevel} union
 * (including `max`, which is session-scoped and deliberately absent from the
 * persisted settings type), every other key keeps its `Settings` type.
 */
export type PendingFlagSettings = {
  [K in keyof Settings]?: K extends "effortLevel"
    ? EffortLevel | null
    : Settings[K] | null;
};

export type SidecarSessionSettings = {
  model: string;
  permissionModeId: string;
  planMode: boolean;
  plansDirectory?: string;
  effort: string;
  speed: string;
};

export type SidecarConfigOption = {
  id: string;
  name?: string;
  description?: string;
  category?: string;
  type?: string;
  currentValue?: string;
  /**
   * Provider-resolved runtime value. This stays separate from currentValue so
   * an inherited `default` selection is not rewritten into a pinned model.
   */
  effectiveValue?: string;
  options: Array<{
    value: string;
    name: string;
    description?: string;
  }>;
};

export function sidecarSessionSettings(
  payload: Record<string, unknown>
): SidecarSessionSettings {
  const settings = recordValue(payload.settings) ?? {};
  const env = recordValue(payload.env) ?? {};
  return {
    model: stringValue(settings.model),
    permissionModeId:
      stringValue(payload.permissionModeId) ||
      stringValue(settings.permissionModeId) ||
      "default",
    planMode: booleanValue(settings.planMode),
    plansDirectory: stringValue(env[claudePlansDirectoryEnv]),
    effort:
      stringValue(payload.effort) ||
      stringValue(settings.effort) ||
      stringValue(settings.reasoningEffort),
    speed: stringValue(settings.speed)
  };
}

export function effectivePermissionMode(
  settings: SidecarSessionSettings
): PermissionMode | undefined {
  if (settings.planMode) {
    return "plan";
  }
  const permissionMode = permissionModeValue(settings.permissionModeId);
  if (permissionMode === "bypassPermissions" && !canBypassPermissions()) {
    return "default";
  }
  return permissionMode;
}

export function permissionModeValue(value: string): PermissionMode | undefined {
  switch (value) {
    case "default":
    case "acceptEdits":
    case "bypassPermissions":
    case "plan":
    case "auto":
      return value;
    case "dontAsk":
      return "default";
    default:
      return undefined;
  }
}

export function modelOptionValue(value: string): string | undefined {
  const model = stringValue(value);
  return model && model !== "default" ? model : undefined;
}

/**
 * Effort levels the CLI/SDK itself understands (`EffortLevel` in the SDK types),
 * i.e. the protocol's ladder. This is not a per-model capability table: the
 * sidecar relays the user's choice and leaves "can this upstream honor it?" to
 * the upstream.
 */
const claudeEffortLevels: readonly EffortLevel[] = [
  "low",
  "medium",
  "high",
  "xhigh",
  "max"
];

export function sidecarModelOptionsFromInitializationResult(
  value: Record<string, unknown>
): SidecarConfigOption["options"] {
  const rawModels = Array.isArray(value.models) ? value.models : [];
  const options: SidecarConfigOption["options"] = [];
  const seen = new Set<string>();
  for (const item of rawModels) {
    const model = recordValue(item);
    if (!model) {
      continue;
    }
    const modelValue =
      stringValue(model.value) ||
      stringValue(model.id) ||
      stringValue(model.modelId) ||
      stringValue(model.model_id);
    if (!modelValue || seen.has(modelValue)) {
      continue;
    }
    seen.add(modelValue);
    const name =
      stringValue(model.displayName) ||
      stringValue(model.display_name) ||
      stringValue(model.name) ||
      modelValue;
    const description = stringValue(model.description);
    options.push({
      value: modelValue,
      name,
      ...(description ? { description } : {})
    });
  }
  return options;
}

export function defaultSidecarModelOptionValue(
  options: SidecarConfigOption["options"]
): string {
  return (
    options.find((option) => option.value === "default")?.value ??
    options[0]?.value ??
    "default"
  );
}

/**
 * Effort level as the live-apply vehicle accepts it: every SDK `EffortLevel`,
 * `max` included (`applyFlagSettings` documents `effortLevel` as additionally
 * accepting `'max'`, session-scoped).
 */
export function effortLevelValue(value: string): EffortLevel | null {
  const level = stringValue(value) as EffortLevel;
  return claudeEffortLevels.includes(level) ? level : null;
}

/**
 * Effort level as create-time `Settings.effortLevel` accepts it. `max` is
 * deliberately excluded by the SDK's persisted settings type ("never persisted
 * to settings files"), so it must ride {@link queryEffortOverride} instead.
 */
export function settingsEffortLevel(
  value: string
): Settings["effortLevel"] | null {
  const level = effortLevelValue(value);
  if (!level || level === "max") {
    return null;
  }
  return level;
}

/**
 * Effort levels create-time settings cannot express and that therefore ride the
 * query `Options.effort` field instead. Currently only `max`; the indirection
 * exists so the two vehicles can never drift apart.
 */
export function queryEffortOverride(value: string): EffortLevel | undefined {
  return effortLevelValue(value) === "max" ? "max" : undefined;
}

/**
 * Whether creating the next query already carries this effort level, i.e.
 * whether live `applyFlagSettings` no longer has to deliver it. Covers both
 * vehicles: `Settings.effortLevel` (low..xhigh) and `Options.effort` (max).
 * Anything else — empty or unknown — is NOT carried and must stay pending, or
 * the level is silently dropped instead of reaching the CLI.
 */
export function queryCreateCarriesEffort(value: string): boolean {
  return (
    settingsEffortLevel(value) !== null ||
    queryEffortOverride(value) !== undefined
  );
}

export function flagSettingsFromSessionSettings(
  settings: SidecarSessionSettings
): PendingFlagSettings {
  const result: PendingFlagSettings = {};
  if (settings.effort) {
    result.effortLevel = effortLevelValue(settings.effort);
  }
  if (settings.speed === "fast") {
    result.fastMode = true;
  } else if (settings.speed === "standard") {
    result.fastMode = false;
  }
  return result;
}

export function querySettingsFromSessionSettings(
  settings: SidecarSessionSettings
): Partial<Settings> {
  const result: Partial<Settings> = {};
  if (settings.plansDirectory) {
    result.plansDirectory = settings.plansDirectory;
  }
  if (settings.speed === "fast") {
    result.fastMode = true;
  } else if (settings.speed === "standard") {
    result.fastMode = false;
  }
  // Bake effort into query create settings. Live applyFlagSettings(effort) on a
  // quiet/resumed Claude query can return while still delaying the next prompt
  // for tens of seconds, which trips host delivery confirmation (~30s).
  const effortLevel = settingsEffortLevel(settings.effort);
  if (effortLevel) {
    result.effortLevel = effortLevel;
  }
  return result;
}

export function approvalOptions(): Array<Record<string, unknown>> {
  return [
    {
      kind: "allow_always",
      name: "Allow for session",
      optionId: "allow_always"
    },
    { kind: "allow_once", name: "Allow", optionId: "allow" },
    { kind: "reject_once", name: "Reject", optionId: "reject" }
  ];
}

export function exitPlanOptions(): Array<Record<string, unknown>> {
  const options = [
    {
      kind: "allow_always",
      name: 'Yes, and use "auto" mode',
      optionId: "auto"
    },
    {
      kind: "allow_always",
      name: "Yes, and auto-accept edits",
      optionId: "acceptEdits"
    },
    {
      kind: "allow_once",
      name: "Yes, and manually approve edits",
      optionId: "default"
    },
    { kind: "reject_once", name: "No, keep planning", optionId: "plan" }
  ];
  if (canBypassPermissions()) {
    options.unshift({
      kind: "allow_always",
      name: "Yes, and bypass permissions",
      optionId: "bypassPermissions"
    });
  }
  return options;
}

export function isAllowOption(optionId: string): boolean {
  return [
    "allow",
    "allow_always",
    "accept",
    "acceptEdits",
    "default",
    "auto",
    "bypassPermissions"
  ].includes(optionId);
}

export function isExitPlanAllowOption(optionId: string): boolean {
  if (optionId === "bypassPermissions") {
    return canBypassPermissions();
  }
  return ["default", "acceptEdits", "auto"].includes(optionId);
}

export function canBypassPermissions(): boolean {
  const isRoot = (process.geteuid?.() ?? process.getuid?.()) === 0;
  return !isRoot || !!process.env.IS_SANDBOX;
}
